chapters/3/modul-kompilujacy.tex

```diff
@@ -0,0 +1,75 @@
+
+\paragraph{Wpływ różnic na wykonanie aplikacji w obu systemach\\}
+Różnice między systemem Linux a Windows są widoczne na każdym aspekcie, począwszy od jądra, systemu plików, wydajności, licencjonowania, dostępnego oprogramowania. Przekładaja się na działanie aplikacji w tych systemach. Komunikacja programu z systemem operacyjnym jest oparta na innych, niekompatybilnych ze sobą interfejsach, oba systemy bazują na innych systemach plików. Aplikacje w systemie Windows korzystają z dynamicznie ładowanych systemowych bibliotek, a swoje ustawienia i konfiguracje często zapisują w rejestrze systemowym. Linuxowe odpowiedniki bibliotek .dll jako pliki .so oraz sposób zapisywania konfiguracji aplikacji w katalogu /etc/ nie są kompatybilne z Windowsem. Sprawia to, funkcjonowanie aplikacji, jak i proces kompilacji oraz wykonania kodu c++ różni się między dwoma systemami.
+
+\subsection{Uruchamianie aplikacji przeznaczonych na system operacyjny Windows na systemie operacyjnym Linux}
+\paragraph{Emulator i warstwa kompatybilności\\}
+Emulator to oprogramowanie tworzące wirtualną maszynę innej platformy. Uruchamianie aplikacji na emulatorze, nie różni się od uruchamiania jej na natywnej platformie, symulowany jest cały system docelowy. Wadą stosowania emulatora jest nieporównywalnie gorsza wydajność w porównaniu do natywnej platformy. W artykule\cite{emulatorPerformance} porównującym wydajność emulatorów Android Studio oraz Blue Stacks do urządzeń fizycznych z systemem Android, uzyskano od trzech, do nawet dwudziestu pięciu razy gorszą wydajność w aplikacji testowej polegającej na wykonaniu algorytmu szachowego mini-max. Warstwa kompatybilności również pozwala na uruchomienie aplikacji zaprojektowanej na inną platformę, jednak zamiast przeprowadzać kosztowną symulację całego systemu docelowego, zamienia ona wołania systemowe pochodzące z uruchamianego programu na te z systemu hosta, dzięki czemu aplikacja działa tak, jak działałaby na natywnej platformie, a system hosta otrzymuje właściwe dla siebie żądania. Dzięki braku konieczności symulowania całej maszyny warstwa kompatybilności zapewnia lepszą wydajność od emulatora i nie wymaga dużego nakładu pamięci.
+
+\paragraph{Wine\\}
+Wine to warstwa kompatybilności aplikacji przeznaczonych do wykonywania na systemie Windows na systemach zgodnych ze standardami POSIX, takich jak Linux, macOS. Jest oprogramowaniem open source, a jego pierwsza wersja została wydana w 1993 roku. Aby uruchomić program przy pomocy Wine należy uruchomić go za pomocą komendy “wine64 <plik wykonywalny>”. Wine ładuje program do pamięci, parsuje go, dołącza potrzebne zależności i przechodzi do wykonywania programu. Gdy w trakcie pracy programu pojawi się wołanie systemowe, Wine przechwytuje je i zamienia na zgodne ze standardami POSIX.
```

Comment:
Wspomniałbym jeszcze o izolacji, czym są WinePrefixy, instalacji dodatkowych bibliotek (winetricks?). Jak już przy tym jesteśmy to można też napisac o forkach wine'a typu Proton?